clippy_lints/src/methods/mod.rs

```diff
@@ -2517,37 +2536,73 @@ fn lint_map_unwrap_or_else<'a, 'tcx>(
     }
 }
 
-/// lint use of `_.map_or(None, _)` for `Option`s
+/// lint use of `_.map_or(None, _)` for `Option`s and `Result`s
 fn lint_map_or_none<'a, 'tcx>(
     cx: &LateContext<'a, 'tcx>,
     expr: &'tcx hir::Expr<'_>,
     map_or_args: &'tcx [hir::Expr<'_>],
 ) {
-    if match_type(cx, cx.tables.expr_ty(&map_or_args[0]), &paths::OPTION) {
-        // check if the first non-self argument to map_or() is None
-        let map_or_arg_is_none = if let hir::ExprKind::Path(ref qpath) = map_or_args[1].kind {
-            match_qpath(qpath, &paths::OPTION_NONE)
-        } else {
-            false
-        };
+    let is_option = match_type(cx, cx.tables.expr_ty(&map_or_args[0]), &paths::OPTION);
+    let is_result = match_type(cx, cx.tables.expr_ty(&map_or_args[0]), &paths::RESULT);
+
+    // There are two variants of this `map_or` lint:
+    // (1) using `map_or` as an adapter from `Result<T,E>` to `Option<T>`
+    // (2) using `map_or` as a combinator instead of `and_then`
+    //
+    // (For this lint) we don't care if any other type calls `map_or`
+    if !is_option && !is_result {
+        return;
+    }
 
-        if map_or_arg_is_none {
-            // lint message
-            let msg = "called `map_or(None, f)` on an `Option` value. This can be done more directly by calling \
-                       `and_then(f)` instead";
-            let map_or_self_snippet = snippet(cx, map_or_args[0].span, "..");
-            let map_or_func_snippet = snippet(cx, map_or_args[2].span, "..");
-            let hint = format!("{0}.and_then({1})", map_or_self_snippet, map_or_func_snippet);
-            span_lint_and_sugg(
-                cx,
-                OPTION_MAP_OR_NONE,
-                expr.span,
-                msg,
-                "try using `and_then` instead",
-                hint,
-                Applicability::MachineApplicable,
-            );
-        }
+    let default_arg_is_none = if let hir::ExprKind::Path(ref qpath) = map_or_args[1].kind {
+        match_qpath(qpath, &paths::OPTION_NONE)
+    } else {
+        false
+    };
+
+    // This is really only needed if `is_result` holds. Computing it here
+    // makes `mess`'s assignment a bit easier, so just compute it here.
+    let f_arg_is_some = if let hir::ExprKind::Path(ref qpath) = map_or_args[2].kind {
+        match_qpath(qpath, &paths::OPTION_SOME)
+    } else {
+        false
+    };
+
+    let mess = if is_option && default_arg_is_none {
```

Comment:
What does `mess` stand for?

I'd prefer destructuring it here and `return;` in the else case.

Reply:
I saw `mess` in a few other lints and assumed it was a convention. I read it as "there is either `Some(mess)` that needs to be linted or no (`None`) mess that needs to be linted". 

Changed to destructure on assignment and return early in d0738bd673fe8e2b42d26b6d116f197f4aecea82